### PR TITLE
New tutorial on how to create a new custom layer in Gluon

### PR DESCRIPTION
**Pull request by Ishitori**
## Description ##
This pull request adds a new tutorial explaining how to create a new custom layer using Gluon API. It doesn't introduce any changes to the code base.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] Code is well-documented

### Changes ###
- New tutorial for Python/Gluon

### CODE REVIEW

**Comment by ThomasDelteil** [comment]
@Ishitori you are missing the download placeholder at the end of your tutorial and you haven't updated the tutorials/index.md file

**Comment by piiswrong** [comment]
We already have one here http://gluon.mxnet.io/chapter03_deep-neural-networks/custom-layer.html
Can we merge these?

File: docs/tutorials/python/custom_layer.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+In this article, I will cover how to create a new layer from scratch, how to use it, what are possible pitfalls and how to avoid them.
+
+## The simplest custom layer
+
+To create a new layer in Gluon API, one must create a class that inherits from [Block](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Block) class. This class provides the most basic functionality, and all predefined layers inherit from it directly or via other subclasses. Because each layer in Apache MxNet inherits from `Block`, words "layer" and "block" are used interchangeably inside of the Apache MxNet community.
+
+The only instance method needed to be implemented is [forward()](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Block.forward), which defines what exactly your layer is going to do during forward propagation. Notice, that it doesn't require to provide what the block should do during backpropagation. Backpropagation pass for blocks is done by Apache MxNet for you.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed these. 
Yes, docs.python.org defines __init__() as a method with a reserved name

File: docs/tutorials/python/custom_layer.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
+
+# How to write a custom layer in Apache MxNet Gluon API
+
+While Gluon API for Apache MxNet comes with [a decent number of predefined layers](https://mxnet.incubator.apache.org/api/python/gluon/nn.html), at some point one may find that a new layer is needed. Adding a new layer in Gluon API is straightforward, yet there are a few things that one needs to keep in mind.
+
+In this article, I will cover how to create a new layer from scratch, how to use it, what are possible pitfalls and how to avoid them.
+
+## The simplest custom layer
+
+To create a new layer in Gluon API, one must create a class that inherits from [Block](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Block) class. This class provides the most basic functionality, and all predefined layers inherit from it directly or via other subclasses. Because each layer in Apache MxNet inherits from `Block`, words "layer" and "block" are used interchangeably inside of the Apache MxNet community.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
+
+# How to write a custom layer in Apache MxNet Gluon API
+
+While Gluon API for Apache MxNet comes with [a decent number of predefined layers](https://mxnet.incubator.apache.org/api/python/gluon/nn.html), at some point one may find that a new layer is needed. Adding a new layer in Gluon API is straightforward, yet there are a few things that one needs to keep in mind.
+
+In this article, I will cover how to create a new layer from scratch, how to use it, what are possible pitfalls and how to avoid them.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Ok

File: docs/tutorials/python/custom_layer.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+```python
+class NormalizationLayer(gluon.Block):
+    def __init__(self):
+        super(NormalizationLayer, self).__init__()
+
+    def forward(self, x):
+        return (x - nd.min(x)) / (nd.max(x) - nd.min(x))
+```
+
+The rest of methods of the `Block` class are already implemented, and majority of them are used to work with parameters of a block. There is one very special method named [hybridize()](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Block.hybridize), though, which I am going to cover before moving to a more complex example of a custom layer.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+## Hybridization and the difference between Block and HybridBlock
+
+Looking into the implementation of [existing layers](https://mxnet.incubator.apache.org/api/python/gluon/nn.html), one may find that more often a block inherits from a [HybridBlock](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.HybridBlock), instead of directly inheriting from `Block` class.
+
+The reason for that is that `HybridBlock` allows to write custom layers that can be used in imperative programming as well as in symbolic programming. It is convinient to support both ways, because of the different values these programming models bring. The imperative programming eases the debugging of the code - one can use regular debugging tools available in modern IDEs to go line by line through the computation. The symbolic programming provides faster execution speed, but harder to debug. You can learn more about the difference between symbolic vs. imperative programming from [this article](https://mxnet.incubator.apache.org/architecture/program_model.html).
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+Because of these reasons it is recommended to develop a new layer using imperative model, but deploy it using symbolic model.
+
+Hybridization is a process that Apache MxNet uses to create a symbolic graph of a forward computation. Optimization of this computational graph allows to increase performance. Once the symbolic graph is created, Apache MxNet caches and reuses it for subsequent computations.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+To simplify support of both imperative and symbolic programming, Apache MxNet introduce the `HybridBlock` class. Compare to the `Block` class, `HybridBlock` already has its [forward()](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.HybridBlock.forward) method implemented, but it defines a [hybrid_forward()](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.HybridBlock.hybrid_forward) method that needs to be implemented.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+From API point of view, the main difference between `forward()` and `hybrid_forward()` is an `F` argument. This argument sometimes is refered as a `backend` in the Apache MxNet community. Depending on if hybridization has been done or not, `F` can refer either to [mxnet.ndarray API](https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html) or [mxnet.symbol API](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html). The former is used for imperative programming, and the latter for symbolic programming.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    def hybrid_forward(self, F, x):
+        return F.broadcast_div(F.broadcast_sub(x, F.min(x)), (F.broadcast_sub(F.max(x), F.min(x))))
+```
+
+Thanks to inheriting from HybridBlock, one can easily do forward pass on a given ndarray, either on CPU or GPU. Notice that we don't call `forward()` or `hybrid_forward()` methods directly.
+
+
+```python
+layer = NormalizationHybridLayer()
+layer(nd.array([1, 2, 3], ctx=mx.cpu()))
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+
+
+    
+    [0.  0.5 1. ]
+    <NDArray 3 @cpu(0)>
+
+
+
+As a rule of thumb, one should always implement custom layers by inheriting from `HybridBlock`. This eaeses the development, and doesn't affect execution speed once hybridization is done.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Reformulate it as: This eases the debugging, though require to think of different backends.

File: docs/tutorials/python/custom_layer.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+It is important to notice that hybridization has nothing to do with computation on GPU. One can train both hybridized and non-hybridized networks on both CPU and GPU, though hybridized networks would work faster. It is hard to say in advance how much faster it is going to be.
+
+## Adding a custom layer to a network
+
+While it is possible, custom layers are rarely used separately. Most often they are used with predefined layers to create a neural network. Output of one layer is used as an input of another layer.
+
+Depending on which class you used as a base one, you can use either [Sequential](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.nn.Sequential) or [HybridSequential](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.nn.HybridSequential) container to form a sequential neural network. By adding layers one by one, one adds dependencies of one layer's input from another layer's output. It is worth noting, that both `Sequential` and `HybridSequential` containers inherit from `Block` and `HybridBlock` respectively.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

File: docs/tutorials/python/custom_layer.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+```python
+net = gluon.nn.HybridSequential()                         # Define a Neural Network as a sequence of hybrid blocks
+with net.name_scope():                                    # Used to disambiguate saving and loading net parameters
+    net.add(Dense(5))                                     # Add Dense layer with 5 neurons
+    net.add(NormalizationHybridLayer())                   # Add our custom layer
+    net.add(Dense(1))                                     # Add Dense layer with 1 neurons
+
+
+net.initialize(mx.init.Xavier(magnitude=2.24))            # Initialize parameters of all layers
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Yes, removed it, so it won't distract from main things.

File: docs/tutorials/python/custom_layer.md [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+     [-0.05046433]
+     [-1.2375476 ]
+     [-0.15506986]]
+    <NDArray 5x1 @cpu(0)>
+
+
+
+## Parameters of a custom layer
+
+Usually, a custom layer is more complicated that the one above. Most of the custom layers have a set of associated parameters, sometimes also referred as weights. This is an internal state of a layer. Most often, these parameters are the ones, that we want to learn during backpropagation step, but sometimes these parameters might be just constants we want to use during forward pass.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed the text.
Yes, scales parameter in the example below represents this case: we don't want to learn scales, just want to do a transformation.

File: docs/tutorials/python/custom_layer.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+
+All parameters of a block are stored and accessed via [ParameterDict](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.ParameterDict) class. This class helps with initialization, updating, saving and loading of the parameters. Each layer can have multiple set of parameters, and all of them can be stored in a single instance of the `ParameterDict` class. On a block level, the instance of the `ParameterDict` class is accessible via `self.params` field, and outside of a block one can access all parameters of the network via [collect_params()](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Block.collect_params) method called on a `container`. `ParameterDict` uses [Parameter](https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#mxnet.gluon.Parameter) class to represent parameters inside of Apache MxNet neural network. If parameter doesn't exist, trying to get a parameter via `self.params` will create it automatically.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Container is not a class, but how the group of objects are called on the main page: https://mxnet.incubator.apache.org/api/python/gluon/gluon.html#containers Replaced it with HybridSequential to make things easier.

Added "collect_params() is a convenience method that iterates through child blocks" to the sentence, and split it in two, for simplification. 

Fixed the self.params.get part.

File: docs/tutorials/python/custom_layer.md [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+class NormalizationHybridLayer(gluon.HybridBlock):
+    def __init__(self, hidden_units, scales):
+        super(NormalizationHybridLayer, self).__init__()
+
+        with self.name_scope():
+            self.weights = self.params.get('weights',
+                                           shape=(hidden_units, 0),
+                                           allow_deferred_init=True)
+
+            self.scales = self.params.get('scales',
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Technically, you can, yes. One benefit of having it as a parameter is an out-of-the-box support for serialization/deserialization.

File: docs/tutorials/python/custom_layer.md [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+                                      shape=scales.shape,
+                                      init=mx.init.Constant(scales.asnumpy().tolist()), # Convert to regular list to make this object serializable
+                                      differentiable=False)
+            
+    def hybrid_forward(self, F, x, weights, scales):
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: One of the magics of MXNet. This way you can be sure that weights and scales going to be proper Symbol if the model is hybridized.

File: docs/tutorials/python/custom_layer.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+def print_params(title, net):
+    """
+    Helper function to print out the state of parameters of NormalizationHybridLayer
+    """
+    print(title)
+    hybridlayer_params = {k: v for k, v in net.collect_params().items() if 'normalizationhybridlayer' in k }
+    
+    for key, value in hybridlayer_params.items():
+        print('{} = {}\n'.format(key, value.data()))
+
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Done

File: docs/tutorials/python/custom_layer.md [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+In the example below, we run one forward and one backward passes to show that `weights` do change while `scales` parameter doesn't change during the training.
+
+
+```python
+def print_params(title, net):
+    """
+    Helper function to print out the state of parameters of NormalizationHybridLayer
+    """
+    print(title)
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Done

File: docs/tutorials/python/custom_layer.md [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    [2.]
+    <NDArray 1 @cpu(0)>
+    
+
+
+As it is seen from the output above, `weights` parameter has been changed by the training and `scales` not.
+
+## Conclusion
+
+One important quality of a Deep learning framework is extensibility. Empowered by flexible abstractions, like `Block` and `HybridBlock`, one can easily extend Apache MxNet functionality to match its needs.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Done

File: docs/tutorials/python/custom_layer.md [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+* `F` backend is used for all calculations.
+* The calculation of dot product is done using `F.FullyConnected()` method instead of `F.dot()` method. The one was chosen over another because the former supports automatic infering shapes of inputs while the latter doesn't. This is important to know, if one doesn't want to hard code all the shapes. The best way to learn what operators supports automatic inference of input shapes at the moment is browsing C++ implementation of operators to see if one uses a method `SHAPE_ASSIGN_CHECK();` for `in_shape`. The output shape is always inferred automatically.
+* `hybrid_forward()` method signature has changed. It accepts two new arguments: `weights` and `scales`.
+
+The last peculiarity is due to support of imperative and symbolic programming by `HybridBlock`. During training phase, parameters are passed to the layer by Apache MxNet framework as additional arguments to the method, because they might need to be converted to `Symbols` depending on if the layer was hybridized. One shouldn't use parameters from the class instance directly or from `self.params.get()` method in `hybrid_forward()`, except to get shapes of parameters.
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Yes, I added a few print statements to show how exactly it looks like + small note explaining it once again.

File: docs/tutorials/python/custom_layer.md [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+                                      differentiable=False)
+            
+    def hybrid_forward(self, F, x, weights, scales):
+        normalized_data = F.broadcast_div(F.broadcast_sub(x, F.min(x)), (F.broadcast_sub(F.max(x), F.min(x))))
+        weighted_data = F.FullyConnected(normalized_data, weights, num_hidden=self.weights.shape[0], no_bias=True)
+        scaled_data = F.broadcast_mul(scales, weighted_data)
+        return scaled_data
+```
+
+In the example above 2 set of parameters are defined:
```
[the remark this answers is not in the document]
Comment (Ishitori) [reply docs/tutorials/python/custom_layer.md]: Fixed

**Comment by Ishitori** [comment]
@piiswrong, while topic is the same, the styles and level of details are different, so it is hard to blend it together. And since they are in different repositories, they can co-exist.
